pages/locations/index.js

```diff
@@ -35,63 +35,71 @@ export default class LocationsPage extends Component {
     }
   }
 
   componentDidMount() {
     document.title = 'Link-SF'
 
-    if (navigator) {
-      navigator.geolocation.getCurrentPosition(this.setCurrentLocation)
-    } else {
-      this.setLocations()
-    }
-
     fetchLocations(20)
       .then(locations => {
-        this.setState({ locations })
+        this.setState({ locations }, this.loadRemaining)
+      })
+  }
+
+  setCurrentLocation = () => {
+    if (navigator) {
+      navigator.geolocation.getCurrentPosition(currentLocation => {
+        this.setState({ currentLocation }, this.addDistances)
       })
+    }
   }
 
-  setCurrentLocation = (currentLocation) => {
-    this.setState({ currentLocation }, this.setLocations)
+  addDistances = () => {
+    const { currentLocation, locations, locationsWithDistance } = this.state
+    const locationSlices = R.splitEvery(25, locations)
+
+    if (!currentLocation) return
+
+    this.calculateDistances(locationSlices, currentLocation)
   }
 
-  setLocations = () => {
-    const { currentLocation } = this.state
-
-    if (currentLocation && this.state.locations.length > 0) {
-      calculateAllDistances(this.state.locations, currentLocation)
-        .then(matrixResponse => {
-          const matrixResponses = matrixResponse.rows[0].elements
-          this.setState({ locations: mergeLocationsAndDistances(this.state.locations, matrixResponse) })
-        })
-    } else if (currentLocation) {
-      fetchLocations(20)
-        .then(locations => {
-          locationsCache = locations
-          return calculateAllDistances(locations, currentLocation)
-        })
-        .then(matrixResponse => {
-          const matrixResponses = matrixResponse.rows[0].elements
-          this.setState({ locations: mergeLocationsAndDistances(locationsCache, matrixResponses) })
-        })
+  calculateDistances = (locationSlices, currentLocation) => {
+    const { locationsWithDistance } = this.state
+    const a = R.splitAt(1, locationSlices)
+    const locations = a[0][0] // this will always only have one element, the array of our working locations
+    const remaining = a[1]
+
+    if (locations) {
+      // we can only call the distance matrix API with 25 elements at a time,
+      // and 100 elements per second, so lets call with 25 elements every 250ms
+      setTimeout(_ => {
+        calculateAllDistances(locations, currentLocation)
+          .then(matrixResponse => {
+            const matrixResponses = matrixResponse.rows[0].elements
+            const updatedLocations = mergeLocationsAndDistances(locations, matrixResponses)
+
+            this.setState({ locationsWithDistance: [...locationsWithDistance, ...updatedLocations] }, _ => {
+              this.calculateDistances(remaining, currentLocation)
+            })
+          })
+      }, 250)
     } else {
-      fetchLocations(20)
-        .then(locations => {
-          this.setState({ locations })
-        })
+      // we should have to do this, but mergeLocationsAndDistances modifies the
```

Comment:
should this be "shouldn't" have to do this? I am not 100% clear what this is stating